### PR TITLE
Reorganise project for testing, add Prometheus metrics endpoint

### PR DESCRIPTION
**Pull request by Limess**
Refactored to allow the servicediscovery struct to receive a graphQL client. This should allow easy simulation of errors, either end-to-end or stubbed, in future.

Added a /metrics Prometheus scrape endpoint, and expose to ecs service discovery. Gracefully exit this server on close along with everything else

### CODE REVIEW

File: cmd/biz-ops-service-discovery/main.go [hunk 1]
```diff
@@ -185,26 +72,61 @@ func main() {
...
 		"directory": directory,
-		"tick": tick.Seconds(),
-		"verbose": verbose,
+		"tick":      tick.Seconds(),
+		"verbose":   verbose,
 	}).Info("Biz-Ops service discovery is running.")
 
 	go func() {
-		if err := writeConfiguration(); err != nil {
+		bizopsDiscovery := servicediscovery.BizOps{
```
Comment (Limess) [inline cmd/biz-ops-service-discovery/main.go]: Dependency injection! Looks like static typing is back on the menu

File: deployments/cloudformation.yml [hunk 2]
```diff
@@ -46,9 +53,12 @@ Resources:
               Value: 60s
             - Name: DIRECTORY
               Value: /mnt/efs/prometheus/service-discovery
+            - Name: PORT
+              Value: !Ref ServiceHttpPort
           DockerLabels:
             com.ft.service-name: !Sub "${ServiceName}-service"
             com.ft.service-region: !Ref "AWS::Region"
+            com.prometheus-ecs-discovery.port: !Ref ServiceHttpPort
```
Comment (Limess) [inline deployments/cloudformation.yml]: Used by ecs-service discovery to discover the dynamic port mapping. Haven't tested this yet but will hopefully lead to a nice discovery file.

File: deployments/cloudformation.yml [hunk 3]
```diff
@@ -26,6 +26,13 @@ Parameters:
     Type: String
     NoEcho: true
 
+  ServiceHttpPort:
+    Description: >
+      The port to route http traffic from FT sources. Not exposed externally via ALB.
+      See https://github.com/Financial-Times/monitoring-aggregation-ecs/blob/master/documentation/SERVICE_DEFINITIONS.md for guidelines
+    Type: Number
+    Default: 12006
```
Comment (lucas42) [inline deployments/cloudformation.yml]: This port is being used by the cloudwatch alarm exporter

File: internal/servicediscovery/servicediscovery.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	healthChecks := responsePayload.Data.Healthchecks
+
+	configuration := make([]prometheusConfiguration, 2)
+
+	configuration[0].Labels = map[string]string{"observe": "yes"}
+	configuration[1].Labels = map[string]string{"observe": "no"}
+
+	for _, healthCheck := range healthChecks {
+		// Check the URL is legit, ignore it on parse errors.
+		_, err := url.Parse(healthCheck.URL)
```
Comment (lucas42) [inline internal/servicediscovery/servicediscovery.go]: I know you've just copied this logic from a different file, but I'm not sure it be should the job of the service discovery app to validate that a healthcheck is a valid URL.  I'd prefer this type of thing to be checked by `healthcheck-validator` and have the exporter handle the error.

File: internal/server/server.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+func Server(listenAddress string) *http.Server {
+	router := http.NewServeMux()
+
+	router.Handle("/metrics", promhttp.Handler())
+
+	logger := logrus.New()
+	w := logger.Writer()
+	defer w.Close()
+
+	server := &http.Server{
```
[the remark this answers is not in the document]
Comment (Limess) [reply internal/server/server.go]: Doesn't really matter here so I've omitted this.

File: deployments/cloudformation.yml (hunk 3 above)
Comment (Limess) [reply deployments/cloudformation.yml]: Nikita has fixed this.

File: internal/servicediscovery/servicediscovery.go (hunk 4 above)
Comment (Limess) [reply internal/servicediscovery/servicediscovery.go]: Let's handle this later when we add it.

**Pull request closed without merging**